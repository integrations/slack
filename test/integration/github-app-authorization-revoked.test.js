const nock = require('nock');
const Sequelize = require('sequelize');


const { probot, models } = require('.');
const githubAppAuthorizationRevoked = require('../fixtures/webhooks/github_app_authorization.revoked.json');

const { Op } = Sequelize;
const {
  SlackWorkspace,
  GitHubUser,
  SlackUser,
  Installation,
  Subscription,
} = models;

describe('Integration: github_app_authorization.revoked', () => {
  let workspace;
  let githubUser;
  let slackUser;
  let installation1;
  let installation2;
  beforeEach(async () => {
    // create workspace
    workspace = await SlackWorkspace.create({
      slackId: 'T0001',
      accessToken: 'xoxp-token',
    });

    githubUser = await GitHubUser.create({
      id: 1,
      accessToken: 'secret',
    });

    slackUser = await SlackUser.create({
      slackId: 'U2147483697',
      slackWorkspaceId: workspace.id,
      githubId: githubUser.id,
    });

    installation1 = await Installation.create({
      githubId: 1,
      ownerId: 1,
    });

    installation2 = await Installation.create({
      githubId: 2,
      ownerId: 2,
    });

    await Subscription.subscribe({
      channelId: 'C2147483705',
      githubId: 1,
      installationId: installation1.id,
      slackWorkspaceId: workspace.id,
      creatorId: slackUser.id,
      type: 'repo',
    });

    await Subscription.subscribe({
      channelId: 'C2147483705',
      githubId: 2,
      installationId: installation2.id,
      slackWorkspaceId: workspace.id,
      creatorId: slackUser.id,
      type: 'repo',
    });

    await Subscription.subscribe({
      channelId: 'C12345',
      githubId: 3,
      installationId: installation2.id,
      slackWorkspaceId: workspace.id,
      creatorId: slackUser.id,
      type: 'repo',
    });
  });

  test('works for a user  who has connected their GitHub account to one Slack workspace', async () => {
    nock('https://api.github.com').get('/repositories/1').reply(200, {
      full_name: 'atom/atom',
    });

    nock('https://api.github.com').get('/repositories/2').reply(200, {
      full_name: 'kubernetes/kubernetes',
    });

    nock('https://api.github.com').get('/repositories/3').reply(200, {
      full_name: 'integrations/slack',
    });

    nock('https://slack.com').post('/api/chat.postMessage').times(2).reply(200, { ok: true });

    await probot.receive({
      name: 'github_app_authorization',
      payload: githubAppAuthorizationRevoked,
    });

    expect((await slackUser.reload()).githubId).toBe(null);
    expect((await Subscription.findAll({ where: { creatorId: slackUser.id } })).length).toBe(0);
  });
  test('works for a user who has connected their GitHub account to multiple Slack workspaces', async () => {
    const workspace2 = await SlackWorkspace.create({
      slackId: 'T0002',
      accessToken: 'xoxp-token2',
    });
    const slackUser2 = await SlackUser.create({
      slackId: 'U123456',
      slackWorkspaceId: workspace.id,
      githubId: githubUser.id,
    });
    await Subscription.subscribe({
      channelId: 'C12121212',
      githubId: 3,
      installationId: installation2.id,
      slackWorkspaceId: workspace2.id,
      creatorId: slackUser2.id,
      type: 'repo',
    });

    await Subscription.subscribe({
      channelId: 'C9999999',
      githubId: 2,
      installationId: installation2.id,
      slackWorkspaceId: workspace2.id,
      creatorId: slackUser2.id,
      type: 'repo',
    });

    nock('https://api.github.com').get('/repositories/1').reply(200, {
      full_name: 'atom/atom',
    });

    nock('https://api.github.com').get('/repositories/2').times(2).reply(200, {
      full_name: 'kubernetes/kubernetes',
    });

    nock('https://api.github.com').get('/repositories/3').times(2).reply(200, {
      full_name: 'integrations/slack',
    });

    nock('https://slack.com').post('/api/chat.postMessage').times(4).reply(200, { ok: true });

    await probot.receive({
      name: 'github_app_authorization',
      payload: githubAppAuthorizationRevoked,
    });

    expect((await slackUser.reload()).githubId).toBe(null);
    expect((await slackUser2.reload()).githubId).toBe(null);
    expect((await Subscription.findAll({
      where: {
        creatorId: {
          [Op.or]: [slackUser.id, slackUser2.id],
        },
      },
    })).length).toBe(0);
  });
});
