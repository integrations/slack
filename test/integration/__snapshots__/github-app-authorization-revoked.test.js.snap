// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Integration: github_app_authorization.revoked works for a user  who has connected their GitHub account to one Slack workspace 1`] = `
Object {
  "attachments": "[{\\"color\\":\\"#24292f\\",\\"mrkdwn_in\\":[\\"text\\"],\\"text\\":\\"Subscriptions to 2 repositories have been disabled, because <@U2147483697>, who originally set them up, has disconnected their GitHub account.\\\\nUse the following slash commands to re-enable the subscriptions:\\\\n/github subscribe atom/atom\\\\n/github subscribe kubernetes/kubernetes\\"}]",
  "channel": "C2147483705",
  "token": "xoxp-token",
}
`;

exports[`Integration: github_app_authorization.revoked works for a user  who has connected their GitHub account to one Slack workspace 2`] = `
Object {
  "attachments": "[{\\"color\\":\\"#24292f\\",\\"mrkdwn_in\\":[\\"text\\"],\\"text\\":\\"The subscription to 1 repository has been disabled, because <@U2147483697>, who originally set it up, has disconnected their GitHub account.\\\\nUse the following slash command to re-enable the subscription:\\\\n/github subscribe integrations/slack\\"}]",
  "channel": "C12345",
  "token": "xoxp-token",
}
`;

exports[`Integration: github_app_authorization.revoked works for a user who has connected their GitHub account to multiple Slack workspaces 1`] = `
Object {
  "attachments": "[{\\"color\\":\\"#24292f\\",\\"mrkdwn_in\\":[\\"text\\"],\\"text\\":\\"Subscriptions to 2 repositories have been disabled, because <@U2147483697>, who originally set them up, has disconnected their GitHub account.\\\\nUse the following slash commands to re-enable the subscriptions:\\\\n/github subscribe atom/atom\\\\n/github subscribe kubernetes/kubernetes\\"}]",
  "channel": "C2147483705",
  "token": "xoxp-token",
}
`;

exports[`Integration: github_app_authorization.revoked works for a user who has connected their GitHub account to multiple Slack workspaces 2`] = `
Object {
  "attachments": "[{\\"color\\":\\"#24292f\\",\\"mrkdwn_in\\":[\\"text\\"],\\"text\\":\\"The subscription to 1 repository has been disabled, because <@U2147483697>, who originally set it up, has disconnected their GitHub account.\\\\nUse the following slash command to re-enable the subscription:\\\\n/github subscribe integrations/slack\\"}]",
  "channel": "C12345",
  "token": "xoxp-token",
}
`;

exports[`Integration: github_app_authorization.revoked works for a user who has connected their GitHub account to multiple Slack workspaces 3`] = `
Object {
  "attachments": "[{\\"color\\":\\"#24292f\\",\\"mrkdwn_in\\":[\\"text\\"],\\"text\\":\\"The subscription to 1 repository has been disabled, because <@U123456>, who originally set it up, has disconnected their GitHub account.\\\\nUse the following slash command to re-enable the subscription:\\\\n/github subscribe integrations/slack\\"}]",
  "channel": "C12121212",
  "token": "xoxp-token",
}
`;

exports[`Integration: github_app_authorization.revoked works for a user who has connected their GitHub account to multiple Slack workspaces 4`] = `
Object {
  "attachments": "[{\\"color\\":\\"#24292f\\",\\"mrkdwn_in\\":[\\"text\\"],\\"text\\":\\"The subscription to 1 repository has been disabled, because <@U123456>, who originally set it up, has disconnected their GitHub account.\\\\nUse the following slash command to re-enable the subscription:\\\\n/github subscribe kubernetes/kubernetes\\"}]",
  "channel": "C9999999",
  "token": "xoxp-token",
}
`;
